app/src/main/java/com/arctouch/codechallenge/view/MovieListFragment.kt

```diff
@@ -0,0 +1,152 @@
+                        adapter.notifyDataSetChanged()
+                        dismissSwipeRefresh()
+                        dismissSnackBar()
+                    }
+                    //loading. show swiperefresh
+                    Resource.Status.LOADING ->{
+                        showSwipeRefresh()
+                    }
+                    //error. show snackbar with error
+                    Resource.Status.ERROR ->{
+                        dismissSwipeRefresh()
+                        showSnackBar(it.throwable)
+                    }
+                }
+            }
+        })
+
+        //get data
+        viewModel.getMoviesData()
+    }
+
+    //Setup recyclerview layout manager and listeners
+    private fun setupRecyclerViewAndListeners(){
+        //get screen orientation and set gridlayout columns
+        val orientation = resources.configuration.orientation
+
+        val layoutManager = if (orientation == Configuration.ORIENTATION_LANDSCAPE)
+            GridLayoutManager(activity, 3)
+         else
+            GridLayoutManager(activity, 2)
+
+        recyclerView.layoutManager = layoutManager
+        //
+
+        //adapter setup
+        adapter = MoviesListAdapter(movies)
+        adapter.setOnItemClickListener(object : MoviesListAdapter.ClickListener{
+            override fun onItemClick(position: Int, v: View, item: com.arctouch.codechallenge.model.Movie) {
+                navigationViewModel.changeFragmentTo(NavigationViewModel.AvailableFragments.MovieDetail, item)
+            }
+        })
+        recyclerView.adapter = adapter
+        //
+
+        //endless scroll listener setup
+        scrollListener = object: EndlessRecyclerViewScrollListener(layoutManager){
+            override fun onLoadMore(page: Int, totalItemsCount: Int, view: RecyclerView?) {
+                viewModel.loadMoreMovies()
+            }
+        }
+        recyclerView.addOnScrollListener(scrollListener)
+        //
+
+        //SwipeToRefresh listener
+        swipeToRefresh.setOnRefreshListener {
+            viewModel.getMoviesData(true)
+            movies.clear()
+            scrollListener.resetState()
+        }
+        //
+    }
+
+    //region swipe refresh controller
+    private fun showSwipeRefresh(){
+        if(!swipeToRefresh.isRefreshing) swipeToRefresh?.isRefreshing = true
+    }
+
+    private fun dismissSwipeRefresh(){
+        if(swipeToRefresh.isRefreshing) swipeToRefresh?.isRefreshing = false
+    }
+    //endregion
+
+    //region snackbar controler
+    private fun showSnackBar(throwable: Throwable? = null){
+        snackbar = if(throwable != null && !throwable.localizedMessage.isNullOrEmpty()){
+            Snackbar.make(activity!!.findViewById(android.R.id.content), throwable.localizedMessage, Snackbar.LENGTH_INDEFINITE)
+        }else{
+            Snackbar.make(activity!!.findViewById(android.R.id.content), getString(R.string.noNetworkAvailable), Snackbar.LENGTH_INDEFINITE)
```

Comment:
When would this be called?

Reply:
it's a helper to show errors to the user, it's not the great way to do that but it's something that i use in this project to show an network error for example.